### PR TITLE
Add all seven of the javafx modules to JavaBuild.java

### LINKED ISSUE #15: Error trying to import JavaFX classes in Processing 4
## Description
I have a project based on Processing 3 and I'm trying to make it work with Processing 4. It uses a few JavaFX classes and I just can't get them to work. I know JavaFX was moved to a separate library but even after installing it I get an error trying to import JavaFX classes in Processing 4.

BTW I asked this same question in [discourse.processing.org](https://discourse.processing.org/t/how-to-use-javafx-in-processing-4/35842) but no one seems to know.

## Expected Behavior
JavaFX imports should work once you install the JavaFX library.

## Current Behavior
Installing the JavaFX library and trying to import a JavaFX class gives me this error: `The package “javafx” does not exist. You might be missing a library.`

## Steps to Reproduce
1. Install JavaFX library in Processing 4.
2. Restart Processing.
3. Try to run this code:
```
import javafx.beans.property.SimpleIntegerProperty;

void setup() {
  SimpleIntegerProperty version = new SimpleIntegerProperty(1);
}

void draw() {
  background(0);
}
```

## Your Environment
* Processing version: Processing 4.0 beta 7
* Operating System and OS version: Windows 10
* Other information: JavaFX library v4.0 beta 5

## Possible Causes / Solutions
If there's anything I need to do after installing the JavaFX library, please let me know.

### LINKED ISSUE #17: Add modules for JavaFX
Currently there are only three out of seven JavaFX modules included in the Processing editor which severely limits our ability to use it.  Please add all seven JavaFX modules to the Processing editor.  I have modified two Processing editors using the technique below and can verify that it works.  The problem is that even if I write demos using the other modules there is only one other user in the forum who can run them to the best of my knowledge (the other user modified his editor also).  One user found that he could run the demo if all the module jar files were copy/pasted into a 'code' folder in the sketch folder.  The modules' jar files may be found in the 'libraries' folder.  I have been unable to verify that this technique works reliably and frequently other users don't want to do all this just to run a demo.  It would be easier for everyone if all the JavaFX modules were added to the Processing editor source code as follows:

1. Start with https://github.com/processing/processing4
2. Open the “java” folder.
3. Open the “src/processing/mode/java” folder.
4. Open the file “JavaBuild.java”
5. Go to line 1086
6. change: “–add-modules”, “javafx.base,javafx.graphics,javafx.swing”,
to: “–add-modules”, “javafx.base,javafx.graphics,javafx.swing,javafx.controls,javafx.media,javafx.web,javafx.fxml”,

Thanks.

### PR DESCRIPTION
**Pull request by vsquared**
Please add _all seven of the javafx modules so that we may use JavaFX controls in our apps. Thanks.

Closes https://github.com/processing/processing4-javafx/issues/17
Closes https://github.com/processing/processing4-javafx/issues/15

### CODE REVIEW

**Comment by hx2A** [comment]
I'd like to second this. This change would also make it much easier to use JavaFX with py5.

**Comment by SableRaf** [comment]
Hi @vsquared and thanks for your suggestion. It looks like the current commit just adds a file to the root of the repository with instructions for modifying `JavaBuild.java`, instead of directly making the change in the file. Could you update the PR to modify `JavaBuild.java` directly instead so we can properly review the proposed changes?

**Comment by glvjr** [comment]
Hello,

I am seeing different behavior with Processing 4.3 and 4.3.1

These will run in Processing 4.3.1 with the JavaFX jar files (All or only required) in the sketch folder:
[https://discourse.processing.org/t/javafx-controls-preview/41353](url)
[https://discourse.processing.org/t/javafx-webview-preview/41343](url)

They (above) will give errors in Processing 4.3

I did a build of Processing 4.3.1 with IntelliJ IDEA 2024.3 (community Edition) with this:

` "--add-modules", "javafx.base,javafx.graphics,javafx.swing,javafx.controls,javafx.media,javafx.web,javafx.fxml",`
And I still had to add the required jar files to the sketch folder.

This is on a W10 PC.

:)

**Comment by vsquared** [comment]
After modifying two different Processing editors with the above change my observation is that adding the seven javafx  jar files to the sketch folder is no longer necessary.  I have no experience using javafx in IntelliJ or testing on a platform other than macos.  My impression is that adding all the jar files to the sketch folder sometimes fails when trying to run javafx code in an editor that has not been modified.  I cannot be 100% certain about this and it needs further study.  Nevertheless, I don't think the average user wants to go to all the bother to find the seven jar files and copy them to the sketch folder just to run a javafx demo.  It would be easier for everyone if a simple one line change was made to the editor.  I know this works for macos, but have not tried it on Windows or Linux.
**Addendum:**
Processing 4.3 failed to run a javafx app with all seven jar files in sketch folder, either lying free or inside a 'code' folder.  Another copy of a Processing 4.3 editor modified with the changes above _will_ run the javafx demo _without_ the addition of jar files.
Processing 4.3.1 on macos also failed to run a javafx app with all seven jar files in the sketch folder, both lying free and inside a 'code' folder.

**Comment by vsquared** [comment]
@SableRaf - I did another fork of the main app and edited the JavaBuild.java file and submitted a second pull request.

File: vsquared-patch-1
Comment (Stefterv) [inline vsquared-patch-1:1]: Please remove this file

File: java/src/processing/mode/java/JavaBuild.java
Comment (Stefterv) [inline java/src/processing/mode/java/JavaBuild.java]: please remove the `change to` line and remove the old line, don't comment it out

**Review by Stefterv — CHANGES_REQUESTED** [review]
Hi @vsquared,

Please change the things I marked, I will now look into the JavaFX examples provided and see if they work

Looking at the context of this change as something specific and limited to existing cases handling JavaFX code I think it would probably be fine. I just need to test and understand your use case

**Comment by Stefterv** [comment]
Tested on windows, does indeed fix the issue you're describing, so happy to merge after the requested changes. 
If you can find a user friendly way, e.g. not just adding the JavaFX jars to the code folder, please let me know. I would like to make that possible.

**Review by Stefterv — APPROVED** [review]

**Comment by Stefterv** [comment]
Cool thanks @vsquared! 

I'll let you hit the merge button @SableRaf

**Comment by vsquared** [comment]
@Stefterv - `If you can find a user friendly way, e.g. not just adding the JavaFX jars to the code folder, please let me know. I would like to make that possible.`
On macos if you make the changes I suggested the user should _not_ need to add the jar files to the sketch folder.

**Comment by vsquared** [comment]
@Stefterv  **Pull Request Followup:** Thank you for adding the pull request changes.  After the last release of the Processing IDE (v. 4.3.2) I have been able to run JavaFX demos on MacOS without adding a 'code' folder to the sketch or modifying/recompiling the editor as I was doing in the past.  Because of this success I have started posting new JavaFX demos in the Forum with the anticipation that other users would be able to run them too.  Unfortunatey this has not been the case for Windows or Linux users, as discussed here: https://discourse.processing.org/t/processing-and-javafx/45692/6 .  They are able to run JavaFX demos, but only after creating a 'code' folder either by drag 'n dropping all seven of the JavaFX jar files onto an open sketch window or manually copy/pasting the jar files which are not easy to find in the 'libraries' folder.  It appears that the Processing IDE correctly installs the JavaFX jar files but _does not know that they are there_ and therefore unable to use them.  It also appears that this is a known problem and is on the TODO list to fix.  I'm willing to help track this down if I know where to start, but I am not a skilled Windows or Linux programmer.  I do have Windows 11 and Linux PopOS systems which are rarely used but could be used for testing.  Do you have any advice on where to start looking to solve this problem?  I appreciate your time and efforts.

**Comment by SableRaf** [comment]
Hi @vsquared. Thanks for your follow-up and for testing on different platforms. 

For the JavaFX compatibility issue specifically, I recommend [opening a new issue](https://github.com/processing/processing4/issues/new/choose), clarifying the current and intended behavior, as well as any proposed solution(s).

Note: Right now, our main focus is the Gradle migration, which is a major effort with very limited capacity (mostly just @Stefterv working on it). Long term, we are looking at moving away from JOGL and abstracting the renderer—this is part of the roadmap (see [issue #881](https://github.com/processing/processing4/issues/881)). If you have any thoughts on that and how JavaFX may support that goal, feel free to contribute to the discussion over there.

Appreciate your patience and contributions!

**Pull request merged**
